mypy/plugin.py

```diff
@@ -379,3 +392,189 @@ def int_pow_callback(ctx: MethodContext) -> Type:
+        info: TypeInfo,
+        method_name: str,
+        args: List[Argument],
+        ret_type: Type,
+        self_type: Type,
+        function_type: Instance) -> None:
+    from mypy.semanal import set_callable_name
+
+    first = [Argument(Var('self'), self_type, None, ARG_POS)]
+    args = first + args
+
+    arg_types = [arg.type_annotation for arg in args]
+    arg_names = [arg.variable.name() for arg in args]
+    arg_kinds = [arg.kind for arg in args]
+    assert None not in arg_types
+    signature = CallableType(cast(List[Type], arg_types), arg_kinds, arg_names,
+                             ret_type, function_type)
+    func = FuncDef(method_name, args, Block([]))
+    func.info = info
+    func.is_class = False
+    func.type = set_callable_name(signature, func)
+    func._fullname = info.fullname() + '.' + method_name
+    info.names[method_name] = SymbolTableNode(MDEF, func)
+
+
+attr_class_makers = {
+    'attr.s',
+    'attr.attrs',
+    'attr.attributes',
+}
+
+attr_attrib_makers = {
+    'attr.ib',
+    'attr.attrib',
+    'attr.attr'
+}
+
+
+def attr_class_maker_callback(ctx: ClassDefContext) -> None:
+    """Add an __init__ method to classes decorated with attr.s."""
+
+    def get_argument(call: CallExpr, name: Optional[str],
+                     num: Optional[int]) -> Optional[Expression]:
+        for i, (attr_name, attr_value) in enumerate(zip(call.arg_names, call.args)):
+            if num is not None and not attr_name and i == num:
+                return attr_value
+            if name and attr_name == name:
+                return attr_value
+        return None
+
+    def get_bool_argument(call: CallExpr, name: str,
+                          default: Optional[bool]) -> Optional[bool]:
+        for arg_name, arg_value in zip(call.arg_names, call.args):
+            if arg_name == name:
+                # TODO: Handle None being returned here.
+                return ctx.api.parse_bool(arg_value)
+        return default
+
+    def called_function(expr: Expression) -> Optional[str]:
+        if isinstance(expr, CallExpr) and isinstance(expr.callee, RefExpr):
+            return expr.callee.fullname
+        return None
+
+    decorator = ctx.reason
+
+    # Default values of attr.s()
+    init = True  # type: Optional[bool]
+    cmp = True  # type: Optional[bool]
+    auto_attribs = False  # type: Optional[bool]
+
+    if isinstance(decorator, CallExpr):
+        # Read call arguments.
+        # TODO handle these returning None (e.g. bool=SOMETHING_ELSE)
+        init = get_bool_argument(decorator, "init", init)
+        cmp = get_bool_argument(decorator, "cmp", cmp)
+        auto_attribs = get_bool_argument(decorator, "auto_attribs",
+                                         auto_attribs)
+
+    if not init and not cmp:
+        # Nothing to add.
+        return
+
+    # To support nested classes we use fullname(). But fullname is <filename>.<variable name>
+    # and named_type() expects only the name.
```

Comment:
I don't understand this comment. Why do you need the self type here? It is normally bound at the call site (i.e. type variables of a generic type are fixed).

Reply:
Oh I see.  `AnyType(TypeOfAny.unannotated)` is what i need here!  Thanks!

Reply:
Oh wait. I need it for 
`def __lt__(self, other: SelfType)`

Reply:
OK, then please use it only in methods that need it. Also please double-check that this reflects the actual runtime semantics/docs.

Reply:
Actually, I found out that if i do `ctx.api.accept(func)` on the FuncDef I create it does a lot of things for me (setting the type of the self arg, being one of them).  Is that a reasonable thing to do?

Reply:
> Is that a reasonable thing to do?

I am not sure it is really needed, but if you find a case where this is necessary then go ahead.